service/matching/matching_engine.go

```diff
@@ -370,61 +370,66 @@ func (e *matchingEngineImpl) getTaskQueuePartitionManager(
 	partition tqid.Partition,
 	create bool,
 	loadCause loadCause,
-) (taskQueuePartitionManager, error) {
-	pm, err := e.getTaskQueuePartitionManagerNoWait(partition, create, loadCause)
-	if err != nil || pm == nil {
-		return nil, err
-	}
-	if err = pm.WaitUntilInitialized(ctx); err != nil {
-		e.unloadTaskQueuePartition(pm, unloadCauseInitError)
-		return nil, err
-	}
-	return pm, nil
-}
+) (retPM taskQueuePartitionManager, retCreated bool, retErr error) {
+	defer func() {
+		if retErr != nil || retPM == nil {
+			return
+		}
+
+		if retErr = retPM.WaitUntilInitialized(ctx); retErr != nil {
+			e.unloadTaskQueuePartition(retPM, unloadCauseInitError)
+		}
+	}()
 
-// Returns taskQueuePartitionManager for a task queue. If not already cached, and create is true, tries
-// to get new range from DB and create one. This does not block for the task queue to be initialized.
-func (e *matchingEngineImpl) getTaskQueuePartitionManagerNoWait(
-	partition tqid.Partition,
-	create bool,
-	loadCause loadCause,
-) (taskQueuePartitionManager, error) {
 	key := partition.Key()
 	e.partitionsLock.RLock()
 	pm, ok := e.partitions[key]
 	e.partitionsLock.RUnlock()
-	if !ok {
-		if !create {
-			return nil, nil
-		}
+	if ok {
+		return pm, false, nil
+	}
 
-		// If it gets here, write lock and check again in case a task queue is created between the two locks
-		e.partitionsLock.Lock()
-		pm, ok = e.partitions[key]
-		if !ok {
-			namespaceEntry, err := e.namespaceRegistry.GetNamespaceByID(partition.NamespaceId())
-			if err != nil {
-				e.partitionsLock.Unlock()
-				return nil, err
-			}
-			nsName := namespaceEntry.Name()
-			tqConfig := newTaskQueueConfig(partition.TaskQueue(), e.config, nsName)
-			tqConfig.loadCause = loadCause
-			userDataManager := newUserDataManager(e.taskManager, e.matchingRawClient, partition, tqConfig, e.logger, e.namespaceRegistry)
-			pm, err = newTaskQueuePartitionManager(e, namespaceEntry, partition, tqConfig, userDataManager)
-			if err != nil {
-				e.partitionsLock.Unlock()
-				return nil, err
-			}
-			e.partitions[key] = pm
-		}
+	if !create {
+		return nil, false, nil
+	}
+
+	namespaceEntry, err := e.namespaceRegistry.GetNamespaceByID(partition.NamespaceId())
+	if err != nil {
+		return nil, false, err
+	}
+	nsName := namespaceEntry.Name()
+	tqConfig := newTaskQueueConfig(partition.TaskQueue(), e.config, nsName)
+	tqConfig.loadCause = loadCause
+	userDataManager := newUserDataManager(e.taskManager, e.matchingRawClient, partition, tqConfig, e.logger, e.namespaceRegistry)
+	newPM, err := newTaskQueuePartitionManager(e, namespaceEntry, partition, tqConfig, userDataManager)
+	if err != nil {
+		return nil, false, err
+	}
+
+	// If it gets here, write lock and check again in case a task queue is created between the two locks
+	e.partitionsLock.Lock()
+	pm, ok = e.partitions[key]
+	if ok {
 		e.partitionsLock.Unlock()
+		return pm, false, nil
```

Comment:
Did you verify that `newTaskQueuePartitionManager` only allocates memory and doesn't start any goroutines or anything that can be leaked (so it's safe to return and let it get collected)?

Reply:
The only goroutine I can find in structures created and used by `matcher.go`. Its created by a call to `poll()` so I am willing to tentatively say we are safe here, since poll would require that we had started successfully.